### PR TITLE
Create user in new sign-up flow

### PR DESCRIPTION
**Pull request by TurnerRiley**
Creates a teacher user at the end of the new teacher sign-up flow (minus the `school_info` associated with the teacher's school if entered, which will be covered in a followup PR). The new flow uses the same endpoints as the old one but has a bit of extra logic to ensure any unset fields are set by the time the user is created. At the end of the `create` endpoint, the user is then signed in so that the new flow doesn't then send the user to the [sign-in](https://studio.code.org/users/sign_in) page.

### Setting all fields
https://github.com/user-attachments/assets/f1d49fce-7c50-4f54-af40-b626eec62100

Adds entry to DB:
allfields

Saves email preference selection:
allfieldsemail

### Only setting required fields (just display name for teachers)
https://github.com/user-attachments/assets/76accc29-41a6-4268-a10c-310b681ca930

Adds entry to database:
nofields

Saves email preference selection (or lack thereof):
nofieldsemail

## Links
Jira ticket: [here](https://codedotorg.atlassian.net/jira/software/c/projects/ACQ/boards/64?assignee=60d62161f65054006980bd71&selectedIssue=ACQ-2286)

## Testing story
Local testing and adding unit tests.

## Follow-up work
- Add `school_info` to user model upon creating a teacher
- Create student user in student flow

### CODE REVIEW

File: apps/src/signUpFlow/FinishTeacherAccount.tsx [hunk 1]
```diff
@@ -10,10 +10,15 @@ import {
 } from '@cdo/apps/componentLibrary/typography';
 import SchoolDataInputs from '@cdo/apps/templates/SchoolDataInputs';
 
+import {navigateToHref} from '../utils';
+
 import locale from './locale';
 import {
-  DISPLAY_NAME_SESSION_KEY,
-  EMAIL_OPT_IN_SESSION_KEY,
```
Comment (TurnerRiley) [inline apps/src/signUpFlow/FinishTeacherAccount.tsx:18]: Realized it probably doesn't make sense to use `sessionStorage` for the variables set on the same page it's being submitted so removing some of those here! (will remove the student-specific ones in the followup PR that creates the student user)

File: dashboard/test/controllers/registrations_controller_test.rb [hunk 2]
```diff
@@ -222,11 +232,13 @@ class RegistrationsControllerTest < ActionController::TestCase
...
   end
 
+  # USING OLD SIGN-UP FLOW [START]
+
   test "teachers go to specified return to url after signing up" do
     session[:user_return_to] = user_return_to = '//test.code.org/the-return-to-url'
 
     assert_creates(User) do
-      post :create, params: {user: @default_params}
+      post :create, params: {user: @default_params.update(user_type: 'teacher', age: '21+', email_preference_opt_in: true)}
```
Comment (TurnerRiley) [inline dashboard/test/controllers/registrations_controller_test.rb:241]: I noticed this test was labeled as being for teachers but used the default params that use student as the user type so I updated that here.

File: dashboard/test/controllers/registrations_controller_test.rb [hunk 3]
```diff
@@ -516,6 +528,306 @@ class RegistrationsControllerTest < ActionController::TestCase
     assert_equal frozen_time, sign_in.sign_in_at
   end
 
+  # USING OLD SIGN-UP FLOW [END]
+
+  # USING NEW SIGN-UP FLOW [START]
```
Comment (TurnerRiley) [inline dashboard/test/controllers/registrations_controller_test.rb:533]: These are simply copy-pasted from the previous tests but using the new default params that include `new_sign_up = true` to ensure that all the registration components still work with the modified logic of the new flow. I'm hoping this'll make cleaning up the tests after we decide on the old/new signup flow easier since we can just delete a whole chunk.

File: dashboard/app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -26,7 +50,36 @@ def new
       user_params[:user_type] ||= session[:default_sign_up_user_type]
       user_params[:email] ||= params[:email]
 
-      @user = User.new_with_session(user_params.permit(:user_type, :email), session)
+      if !!params[:new_sign_up]
```
Comment (nicklathe) [inline dashboard/app/controllers/registrations_controller.rb]: Slight nit, using `.present?` is preferred to check for the value of this parameter as it's more readable, i.e. `if params[:new_sign_up].present?`

This method checks for nil and empty string values, so this covers the case where the param key may be present but it's value is an empty string.

File: dashboard/app/controllers/registrations_controller.rb [hunk 5]
```diff
@@ -60,10 +113,14 @@ def begin_sign_up
       PartialRegistration.persist_attributes(session, @user)
     end
 
-    render 'new'
+    unless params[:new_sign_up]
```
Comment (nicklathe) [inline dashboard/app/controllers/registrations_controller.rb]: Same here, would be preferable to use `.present?`

File: dashboard/app/controllers/registrations_controller.rb [hunk 6]
```diff
@@ -170,37 +229,41 @@ def create
...
+        user: curr_user,
         event_name: 'Sign Up Finished Backend',
         metadata: event_metadata,
         get_enabled_experiments: true,
       )
     end
 
     SignUpTracking.log_sign_up_result resource, session
+
+    if !!params[:new_sign_up]
```
Comment (nicklathe) [inline dashboard/app/controllers/registrations_controller.rb]: `.present?` here too :)

File: dashboard/app/controllers/registrations_controller.rb [hunk 7]
```diff
@@ -154,10 +211,12 @@ def create
       super
     end
 
-    if current_user && current_user.errors.blank?
-      if current_user.teacher?
+    curr_user = !!params[:new_sign_up] ? User.find_by_email(params[:user_email]) : current_user
```
Comment (nicklathe) [inline dashboard/app/controllers/registrations_controller.rb]: `User.find_by_email` won't work for students, as student emails are hashed. As a result, curr_user will be nil for any student users. Instead you can use `User.find_by_email_or_hashed_email`.

Comment (nicklathe) [reply dashboard/app/controllers/registrations_controller.rb]: I'm probably missing something, but why do we need to to set `curr_user` rather than just use the value of `current_user`, which is set by Devise (by calling super on line 211)?

**Review by nicklathe — COMMENTED** [review]
Hey @TurnerRiley, I left a few comments. I'm going to try testing this branch locally with some of the LTI signup/login flows as well.

File: apps/src/signUpFlow/FinishTeacherAccount.tsx [hunk 8]
```diff
@@ -37,13 +41,39 @@ const FinishTeacherAccount: React.FunctionComponent<{
...
 
-  const onEmailOptInChange = (): void => {
-    const newOptInCheckedChoice = !emailOptInChecked;
-    setEmailOptInChecked(newOptInCheckedChoice);
-    sessionStorage.setItem(
-      EMAIL_OPT_IN_SESSION_KEY,
-      `${newOptInCheckedChoice}`
-    );
+  const submitTeacherAccount = () => {
+    $.ajax({
```
Comment (bethanyaconnor) [inline apps/src/signUpFlow/FinishTeacherAccount.tsx]: nit: use fetch over ajax

File: apps/src/signUpFlow/FinishTeacherAccount.tsx [hunk 9]
```diff
@@ -93,7 +123,7 @@ const FinishTeacherAccount: React.FunctionComponent<{
           className={style.finishSignUpButton}
           color={buttonColors.purple}
           type="primary"
-          onClick={() => console.log('FINISH SIGN UP')}
+          onClick={submitTeacherAccount}
```
Comment (bethanyaconnor) [inline apps/src/signUpFlow/FinishTeacherAccount.tsx:141]: fyi this will have a merge conflict with #61090 (cc @hannahbergam)

File: apps/src/signUpFlow/LoginTypeSelection.tsx [hunk 10]
```diff
@@ -115,16 +123,30 @@ const LoginTypeSelection: React.FunctionComponent = () => {
...
       setEmailIconClass(style.lightGray);
     }
   };
 
-  const finishAccountUrl =
-    sessionStorage.getItem(ACCOUNT_TYPE_SESSION_KEY) === 'teacher'
-      ? studio('/users/new_sign_up/finish_teacher_account')
-      : studio('/users/new_sign_up/finish_student_account');
+  const submitLoginType = () => {
+    $.ajax({
```
Comment (bethanyaconnor) [inline apps/src/signUpFlow/LoginTypeSelection.tsx]: nit: use fetch over ajax

File: apps/test/unit/signUpFlow/FinishStudentAccountTest.tsx [hunk 11]
```diff
@@ -87,24 +86,6 @@ describe('FinishStudentAccount', () => {
     screen.getByText(locale.go_to_my_account());
   });
 
-  it('does not render state question if user is not detected in the U.S.', () => {
```
Comment (bethanyaconnor) [inline apps/test/unit/signUpFlow/FinishStudentAccountTest.tsx]: Why is this removed?

File: dashboard/app/controllers/registrations_controller.rb [hunk 12]
```diff
@@ -14,7 +14,31 @@ class RegistrationsController < Devise::RegistrationsController
     :migrate_to_multi_auth, :demigrate_from_multi_auth
   ]
   skip_before_action :verify_authenticity_token, only: [:set_student_information]
-  skip_before_action :clear_sign_up_session_vars, only: [:new, :begin_sign_up, :cancel, :create]
+  skip_before_action :clear_sign_up_session_vars, only: [:new, :begin_sign_up, :begin_creating_user, :cancel, :create]
+
+  NEW_USER_PERMITTED_PARAMS = [
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/registrations_controller.rb]: can we move this list into a private method such as [`upgrade_params`](https://github.com/code-dot-org/code-dot-org/blob/feb25cc904207409f5ee157df004a7afb445c15c/dashboard/app/controllers/registrations_controller.rb#L608)?

File: dashboard/app/controllers/registrations_controller.rb (hunk 5 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/registrations_controller.rb]: Is `.present?` checking the same thing? This is a boolean, right?

File: dashboard/app/controllers/registrations_controller.rb (hunk 4 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/registrations_controller.rb]: But this is a boolean, right? What does `.present?` return if the value is false?

Comment (TurnerRiley) [reply dashboard/app/controllers/registrations_controller.rb]: I believe `.present?` returns `true` for `false`. `new_sign_up` is just a boolean flag to distinguish between the old and new flows easily

Comment (carl-codeorg) [reply dashboard/app/controllers/registrations_controller.rb]: Nope, it'll return false:
```
false.present?
=> false
```

File: dashboard/app/controllers/registrations_controller.rb (hunk 7 above)
Comment (TurnerRiley) [reply dashboard/app/controllers/registrations_controller.rb]: Oh I was noticing that by that point, the user was considered signed-in (and would have `current_user` set) for the old sign-up flow but would not for the new one, so I was ensuring a user was found for the remaining logic. But actually I think it may make more sense to ensure the user is signed in here then just use `current_user` like you said!

File: dashboard/app/controllers/registrations_controller.rb (hunk 4 above)
Comment (TurnerRiley) [reply dashboard/app/controllers/registrations_controller.rb]: Oh sweet thank you! My bad!

File: apps/test/unit/signUpFlow/FinishStudentAccountTest.tsx (hunk 11 above)
Comment (TurnerRiley) [reply apps/test/unit/signUpFlow/FinishStudentAccountTest.tsx]: Oh whoops I think a merge conflict caused this! Thanks for catching I'll put this back in!

File: dashboard/app/controllers/registrations_controller.rb (hunk 4 above)
Comment (nicklathe) [reply dashboard/app/controllers/registrations_controller.rb]: I find myself referencing this article often when I'm not sure https://blog.appsignal.com/2018/09/11/differences-between-nil-empty-blank-and-present.html

File: dashboard/app/controllers/registrations_controller.rb [hunk 13]
```diff
@@ -542,6 +582,32 @@ def edit
       params[:user][:password].present?
   end
 
+  private def new_user_permitted_params
```
Comment (nicklathe) [inline dashboard/app/controllers/registrations_controller.rb]: I think this would actually be a great candidate to move into [Policies](https://github.com/code-dot-org/code-dot-org/tree/staging/dashboard/lib/policies). There isn't a `registration` policy yet, but the desire is to move stuff like this, that doesn't _need_ to exist in a controller or model, into Policies. This helps us modularize things and mitigate controller and model bloat.

Checkout how we've implemented something very similar in the [lti policy](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/policies/lti.rb#L50).

Then from the context of this controller, to you would just do something like `Policies::Registration::NEW_USER_PERMITTED_PARAMS` if this were a variable, or `Policies::Registration.new_user_permitted_params` if it were a method.

File: apps/src/signUpFlow/FinishTeacherAccount.tsx [hunk 14]
```diff
@@ -39,13 +44,32 @@ const FinishTeacherAccount: React.FunctionComponent<{
...
-      EMAIL_OPT_IN_SESSION_KEY,
-      `${newOptInCheckedChoice}`
-    );
+  const submitTeacherAccount = () => {
+    sendFinishEvent();
+
+    const userType = sessionStorage.getItem(ACCOUNT_TYPE_SESSION_KEY);
+    const email = sessionStorage.getItem(EMAIL_SESSION_KEY);
+    const school = sessionStorage.getItem(SCHOOL_ID_SESSION_KEY);
+    const schoolZip = sessionStorage.getItem(SCHOOL_ZIP_SESSION_KEY);
```
Comment (hannahbergam) [inline apps/src/signUpFlow/FinishTeacherAccount.tsx]: Will we also need a field for country here? I know we're going to tackle the school_infos creation in a followup pr but thinking ahead seeing the school and zip fields accounted for.

**Review by hannahbergam — COMMENTED** [review]
Let me know if there are any weird merge conflicts with the other work we've been doing in this area

File: apps/src/signUpFlow/FinishTeacherAccount.tsx (hunk 14 above)
Comment (TurnerRiley) [reply apps/src/signUpFlow/FinishTeacherAccount.tsx]: Oh good point! I'll add that in thank you

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/app/controllers/registrations_controller.rb [hunk 15]
```diff
@@ -14,19 +14,50 @@ class RegistrationsController < Devise::RegistrationsController
...
+          user_params[:parent_email_preference_source] = EmailPreference::ACCOUNT_SIGN_UP
+        end
+
+        user_params[:data_transfer_agreement_accepted] = params[:data_transfer_agreement_accepted] == '1'
+        if user_params[:data_transfer_agreement_required] && params[:data_transfer_agreement_accepted]
+          user_params[:data_transfer_agreement_accepted] = true
+          user_params[:data_transfer_agreement_request_ip] = request.ip
+          user_params[:data_transfer_agreement_source] = User::ACCOUNT_SIGN_UP
+          user_params[:data_transfer_agreement_kind] = '0'
+          user_params[:data_transfer_agreement_at] = DateTime.now
```
Comment (carl-codeorg) [inline dashboard/app/controllers/registrations_controller.rb]: We're doing quite a bit of param-setting logic now. Do you think it would be worth moving into its own function? It seems like this would clean up the `new` route quite a bit. The function could know about default values for the user params (could store these in a policy), you could pass in any necessary params, and it would return the `user_params`.

File: apps/src/signUpFlow/FinishTeacherAccount.tsx [hunk 16]
```diff
@@ -39,13 +43,31 @@ const FinishTeacherAccount: React.FunctionComponent<{
...
+  const submitTeacherAccount = () => {
+    sendFinishEvent();
+
+    const email = sessionStorage.getItem(EMAIL_SESSION_KEY);
+    const school = sessionStorage.getItem(SCHOOL_ID_SESSION_KEY);
+    const schoolZip = sessionStorage.getItem(SCHOOL_ZIP_SESSION_KEY);
+    const schoolName = sessionStorage.getItem(SCHOOL_NAME_SESSION_KEY);
+    const fetchNewUrlParams = `new_sign_up=true&user_type=teacher&email=${email}&name=${name}&email_preference_opt_in=${emailOptInChecked}&school=${school}&school_id=${school}&school_zip=${schoolZip}&school_name=${schoolName}`;
+
+    fetch(`/users/sign_up?${fetchNewUrlParams}`, {
```
Comment (carl-codeorg) [inline apps/src/signUpFlow/FinishTeacherAccount.tsx]: nitpick: this could be cleaned up by making the outer function async, then awaiting each fetch and calling navigateToHref. It'd eliminate all the promise-chaining boilerplate.

File: dashboard/app/controllers/registrations_controller.rb (hunk 15 above)
Comment (carl-codeorg) [reply dashboard/app/controllers/registrations_controller.rb]: After consulting with @artem-vavilov it could be even better to do a separate [service object](https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1) to encapsulate the business logic from the `if PartialRegistration.in_progress?(session)` block here. I think it's a good opportunity to leave the registrations controller better than we found it. Artem created a [Services::Base](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/services/base.rb) class that also lets you call `:call` directly instead of having to call `new` each time.

Comment (artem-vavilov) [reply dashboard/app/controllers/registrations_controller.rb]: I recommend creating two distinct Service Objects:
- `@user = Services::PartialRegistration::NewUserBuilder.call(request)` for this part:
https://github.com/code-dot-org/code-dot-org/blob/439f03b6e555f28edbdc067d18578bf8877b4635/dashboard/app/controllers/registrations_controller.rb#L26-L54
- `@user = Services::PartialRegistration::UserBuilder.call(request)` for this part:
https://github.com/code-dot-org/code-dot-org/blob/439f03b6e555f28edbdc067d18578bf8877b4635/dashboard/app/controllers/registrations_controller.rb#L56-L59

Comment (TurnerRiley) [reply dashboard/app/controllers/registrations_controller.rb]: Since we intend to only keep one set of logic after we complete the A/B test of the old signup flow vs. new signup flow, do you think it would make sense to only move the new one to a service object?

Comment (artem-vavilov) [reply dashboard/app/controllers/registrations_controller.rb]: Certainly!

File: apps/src/signUpFlow/FinishTeacherAccount.tsx [hunk 17]
```diff
@@ -39,13 +44,35 @@ const FinishTeacherAccount: React.FunctionComponent<{
...
+    const schoolName = sessionStorage.getItem(SCHOOL_NAME_SESSION_KEY);
+    const schoolCountry = sessionStorage.getItem(SCHOOL_COUNTRY_SESSION_KEY);
+    const fetchNewUrlParams = `new_sign_up=true&user_type=teacher&email=${encodedEmail}&name=${name}&email_preference_opt_in=${emailOptInChecked}&school=${school}&school_id=${school}&school_zip=${schoolZip}&school_name=${schoolName}&school_country=${schoolCountry}`;
+
+    await fetch(`/users/sign_up?${fetchNewUrlParams}`, {
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+    });
```
Comment (artem-vavilov) [inline apps/src/signUpFlow/FinishTeacherAccount.tsx]:
```suggestion
    const signUpParams = {
      new_sign_up: true,
      user: {
        user_type: UserTypes.TEACHER,
        email: encodedEmail,
        name: name,
        email_preference_opt_in: emailOptInChecked,
        school: school,
        school_id: school,
        school_zip: schoolZip,
        school_name: schoolName,
        school_country: schoolCountry
      }
    } 

    await fetch(`/users/sign_up?${fetchNewUrlParams}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(signUpParams),
    });
```

File: dashboard/lib/services/partial_registration/new_user_builder.rb [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+          new_user_params[:data_transfer_agreement_source] = User::ACCOUNT_SIGN_UP
+          new_user_params[:data_transfer_agreement_kind] = '0'
+          new_user_params[:data_transfer_agreement_at] = DateTime.now
+        end
+
+        @user = ::User.new_with_session(new_user_params.permit(Policies::Registration::NEW_USER_PERMITTED_PARAMS), @session)
+        @user.save!
+        @user
+      end
+    end
```
Comment (artem-vavilov) [inline dashboard/lib/services/partial_registration/new_user_builder.rb]:
```suggestion
    class UserBuilder < Services::Base
      attr_reader :request

      delegate :session, :params, to: :request

      def initialize(request:)
        @request = request
      end

      def call
        user = ::User.new_with_session(user_params, session)
        user.save!
        user
      end

      private def user_params
        user_params = params.require(:user).permit(*Policies::Registration::NEW_USER_PERMITTED_PARAMS).to_h

        user_params[:user_type] = user_params[:user_type].presence || session[:default_sign_up_user_type]
        case user_params[:user_type]
        when ::User::TYPE_TEACHER
          user_params[:age] = '21+'
          user_params[:email_preference_opt_in_required] = true
          user_params[:email_preference_opt_in] = user_params[:email_preference_opt_in] == 'true' ? 'yes' : 'no'
          user_params[:email_preference_request_ip] = request.ip
          user_params[:email_preference_source] = EmailPreference::ACCOUNT_SIGN_UP
          user_params[:email_preference_form_kind] = '0'
        when ::User::TYPE_STUDENT
          user_params[:parent_email_preference_request_ip] = request.ip
          user_params[:parent_email_preference_source] = EmailPreference::ACCOUNT_SIGN_UP
        end

        user_params[:data_transfer_agreement_accepted] = user_params[:data_transfer_agreement_accepted] == '1'
        if user_params[:data_transfer_agreement_required] && user_params[:data_transfer_agreement_accepted]
          user_params[:data_transfer_agreement_request_ip] = request.ip
          user_params[:data_transfer_agreement_source] = User::ACCOUNT_SIGN_UP
          user_params[:data_transfer_agreement_kind] = '0'
          user_params[:data_transfer_agreement_at] = DateTime.now
        end

        user_params
      end
    end
```

File: dashboard/app/controllers/registrations_controller.rb [hunk 19]
```diff
@@ -14,19 +14,22 @@ class RegistrationsController < Devise::RegistrationsController
...
   def new
     session[:user_return_to] ||= params[:user_return_to]
     if PartialRegistration.in_progress?(session)
-      user_params = params[:user] || ActionController::Parameters.new
-      user_params[:user_type] ||= session[:default_sign_up_user_type]
-      user_params[:email] ||= params[:email]
-
-      @user = User.new_with_session(user_params.permit(:user_type, :email), session)
+      if params[:new_sign_up].present?
+        Services::PartialRegistration::NewUserBuilder.call(request: request)
```
Comment (artem-vavilov) [inline dashboard/app/controllers/registrations_controller.rb]:
```suggestion
        @user = Services::PartialRegistration::UserBuilder.call(request: request)
```

File: apps/src/signUpFlow/LoginTypeSelection.tsx [hunk 20]
```diff
@@ -109,12 +117,35 @@ const LoginTypeSelection: React.FunctionComponent = () => {
...
   };
 
+  const submitLoginType = async () => {
+    logUserLoginType('email');
+
+    const submitLoginTypeParams = {
+      new_sign_up: true,
+      email: email,
+      password: password,
+      password_confirmation: password,
```
Comment (artem-vavilov) [inline apps/src/signUpFlow/LoginTypeSelection.tsx]:
```suggestion
      user: {
        email: email,
        password: password,
        password_confirmation: password,
      },
```

File: dashboard/app/controllers/registrations_controller.rb [hunk 21]
```diff
@@ -52,18 +51,22 @@ def save_default_sign_up_user_type
   # Submit step 1 of the signup process for creating an email/password account.
   #
   def begin_sign_up
-    @user = User.new(begin_sign_up_params)
+    @user = User.new(params[:new_sign_up].present? ? params.permit(:email, :password, :password_confirmation) : begin_sign_up_params)
```
Comment (artem-vavilov) [inline dashboard/app/controllers/registrations_controller.rb]:
```suggestion
    @user = User.new(begin_sign_up_params)
```

File: apps/test/unit/signUpFlow/LoginTypeSelectionTest.tsx [hunk 22]
```diff
@@ -75,59 +92,171 @@ describe('LoginTypeSelection', () => {
...
+    const emailInput = screen.getByLabelText(locale.email_address());
+    const passwordInput = screen.getByLabelText(locale.password());
+    const confirmPasswordInput = screen.getByLabelText(
+      locale.confirm_password()
+    );
+    const beginSignUpParams = {
+      new_sign_up: true,
+      email: email,
+      password: password,
+      password_confirmation: password,
```
Comment (artem-vavilov) [inline apps/test/unit/signUpFlow/LoginTypeSelectionTest.tsx]:
```suggestion
      user: {
        email: email,
        password: password,
        password_confirmation: password,
      },
```

File: dashboard/lib/policies/registration.rb [hunk 23]
```diff
@@ -0,0 +1,25 @@
+class Policies::Registration
```
Comment (artem-vavilov) [inline dashboard/lib/policies/registration.rb]:
```suggestion
module Policies::Registration
```

File: dashboard/lib/policies/registration.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+  NEW_USER_PERMITTED_PARAMS = [
```
Comment (artem-vavilov) [inline dashboard/lib/policies/registration.rb]: I'm not sure we need a separate constant for these parameters.
It should be sufficient to move them directly into `Services::PartialRegistration::UserBuilder#user_params`

File: dashboard/app/controllers/registrations_controller.rb [hunk 25]
```diff
@@ -154,6 +157,12 @@ def create
       super
     end
 
+    if params[:new_sign_up].present?
+      session[:user_return_to] ||= params[:user_return_to]
+      @user = Services::PartialRegistration::UserBuilder.call(request: request)
```
Comment (carl-codeorg) [inline dashboard/app/controllers/registrations_controller.rb:162]: This looks a lot cleaner now. Thanks for implementing the new service!

File: dashboard/test/controllers/registrations_controller_test.rb (hunk 3 above)
Comment (carl-codeorg) [reply dashboard/test/controllers/registrations_controller_test.rb:533]: > after we decide on the old/new signup flow

Is there actually a chance that we decide to keep the old flow? Asking because if we are pretty much for sure refactoring the flow, it might make sense to move the new tests over to the [spec syntax](https://www.betterspecs.org/) now while we're touching the test files. It's pretty much plug-and-play and you can add tests to an old file without changing anything in existing tests: [example](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/controllers/lti/v1/account_linking_controller_test.rb#L118-L163).

**Comment by TurnerRiley** [comment]
@carl-codeorg Sounds good! I'm merging now because this PR is blocking a lot of others, but I'll track that work in a followup!
